File: src/com/ds/array/practice/SpiralMatrix.java
package com.ds.array.practice;

/**
 * Problem seen in FAMG interviews
 *
 * Given 2D (N x M) array, re-arrange / print elements in spiral order
 *
 * e.g.
 *  [-] [-] [-] [->]
 *  [|][-][->][|]
 *  [|][<-][|][|]
 *  [<-][-][-][|]
 */

public class SpiralMatrix {
  static void printArr(int[][] arr) {
    int firstRow = 0,
        lastRow = arr.length - 1,
        firstCol = 0,
        lastCol = arr[0].length - 1;
    
    while (firstRow < lastRow && firstCol < lastCol) {
  
      for (int i = firstCol; i <= lastCol; i++) {
        System.out.print(arr[firstRow][i] + " ");
      }
      for (int i = firstRow + 1; i <= lastRow; i++) {
        System.out.print(arr[i][lastCol] + " ");
      }
      for (int i = lastCol - 1; i > firstCol; i--) {
        System.out.print(arr[lastRow][i] + " ");
      }
      for (int i = lastRow - 1; i > firstCol; i--) {
        System.out.print(arr[i][firstCol] + " ");
      }
      
      firstRow++;
      lastRow--;
      firstCol++;
      lastCol--;
    }
  }
  
  public static void main(String[] args) {
    int arr[][] = {
      {1, 2, 3, 4},
      {5, 6, 7, 8},
      {9, 10, 11, 12},
      {13, 14, 15, 16}
    };
    printArr(arr);
  }
}

File: src/js/arrays/SpiralMatrix.js
/**
 * Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.
 *
 * Input:
 * [
 *   [ 1, 2, 3 ],
 *   [ 4, 5, 6 ],
 *   [ 7, 8, 9 ]
 * ]
 * Output: [1,2,3,6,9,8,7,4,5]
 *
 * Source: https://leetcode.com/problems/spiral-matrix/
 */

function spiralOrder(matrix) {
  let result = [];
  let firstRow = 0,
      lastRow = matrix.length - 1,
      firstCol = 0,
      lastCol = matrix[0].length - 1;

  while (firstRow <= lastRow && firstCol <= lastCol) {

    for (let i = firstCol; i <= lastCol; i++) result.push(matrix[firstRow][i]);
    for (let i = firstRow + 1; i <= lastRow; i++) result.push(matrix[i][lastCol]);

    if (firstRow < lastRow && firstCol < lastCol) {
      for (let i = lastCol - 1; i > firstCol; i--) result.push(matrix[lastRow][i]);
      for (let i = lastRow; i > firstRow; i--) result.push(matrix[i][firstCol]);
    }

    firstRow++;
    lastRow--;
    firstCol++;
    lastCol--;
  }

  return result;
}

const matrix1 = [
  [ 1, 2, 3 ],
  [ 4, 5, 6 ],
  [ 7, 8, 9 ]
];

const matrix2 = [
  [1, 2, 3, 4],
  [5, 6, 7, 8],
  [9,10,11,12]
];

console.log( spiralOrder(matrix1) );
console.log( spiralOrder(matrix2) );
